src/main/ruby/truffleruby/core/string.rb

```diff
@@ -1097,6 +1097,92 @@ def rindex(sub, finish = undefined)
     nil
   end
 
+  def byteindex(str, start = undefined)
+    is_regex_pattern = Primitive.is_a?(str, Regexp)
+
+    if Primitive.undefined?(start)
+      start = 0
+    else
+      start = Primitive.rb_to_int(start)
+
+      start += bytesize if start < 0
+      if start < 0 || start > bytesize
+        if is_regex_pattern
+          Primitive.regexp_last_match_set(Primitive.caller_special_variables, nil)
+        end
+
+        return
+      end
+    end
+
+    unless on_codepoint_boundary?(start)
+      raise(IndexError, "offset #{start} does not land on character boundary")
+    end
+
+    if is_regex_pattern
+      Primitive.encoding_ensure_compatible(self, str)
+
+      match = Truffle::RegexpOperations.match_from(str, self, start)
+      Primitive.regexp_last_match_set(Primitive.caller_special_variables, match)
+      return match ? Primitive.character_index_to_byte_index(self, match.begin(0)) : nil
+    end
+
+    str = StringValue(str)
+    return start if str == ''
+    return if start + str.bytesize > bytesize
+
+    enc = Primitive.encoding_ensure_compatible_str(self, str)
+    Primitive.string_byte_index(self, str, enc, start)
+  end
+
+  def byterindex(str, finish = undefined)
+    if Primitive.undefined?(finish)
+      finish = bytesize
+    else
+      finish = Primitive.rb_to_int(finish)
+      finish += bytesize if finish < 0
+      return if finish < 0
+
+      finish = bytesize if finish > bytesize
+    end
+
+    unless on_codepoint_boundary?(finish)
+      raise(IndexError, "offset #{finish} does not land on character boundary")
+    end
+
+    if Primitive.is_a?(str, Regexp)
+      Primitive.encoding_ensure_compatible(self, str)
+
+      match = Truffle::RegexpOperations.search_region(str, self, 0, finish, false, true)
+      Primitive.regexp_last_match_set(Primitive.caller_special_variables, match)
+      return match ? Primitive.character_index_to_byte_index(self, match.begin(0)) : nil
+    end
+
+    str = StringValue(str)
+    return finish if str.empty?
+
+    # Add `str.size` worth of bytes to `finish` to compensate for `LastByteIndexOfStringNode` which does a reverse
+    # search including the full pattern length (as start pos).
+    finish_adjusted = Primitive.byte_index_to_character_index(self, finish)
+    finish_adjusted += str.size
+    finish_adjusted = size if finish_adjusted > size
+    finish = Primitive.character_index_to_byte_index(self, finish_adjusted)
+
+    enc = Primitive.encoding_ensure_compatible_str(self, str)
+    Primitive.string_byte_reverse_index(self, str, enc, finish)
+  end
+
+  private def on_codepoint_boundary?(byte_pos)
+    char_pos = Primitive.byte_index_to_character_index(self, byte_pos)
+    adjusted_byte_pos = if char_pos >= size
+                          bytesize
+                        else
+                          Primitive.character_index_to_byte_index(self, char_pos)
+                        end
+
+    byte_pos == adjusted_byte_pos
+  end
```

Comment:
**note:** this method should be moved to the `StringOperations` module before merging

Reply:
Done